akka-persistence-testkit/src/test/scala/akka/persistence/testkit/scaladsl/UnpersistentDurableStateSpec.scala

```diff
@@ -0,0 +1,257 @@
+      def persistAdd[Reply](n: Int, replyTo: RecipientRef[Reply], reply: Reply): Effect[State] = {
+        val newState = state.processAdd(n)
+
+        Effect
+          .persist(newState)
+          .thenRun { nextState => // should be the same as newState, but...
+            state.notifyAfter.keysIterator
+              .filter { at =>
+                (at <= nextState.nextNotifyAt) && !(nextState.notifyAfter.isDefinedAt(at))
+              }
+              .foreach { at =>
+                state.notifyAfter(at) ! Done
+              }
+
+            replyTo ! reply
+          }
+          .thenUnstashAll()
+      }
+
+      cmd match {
+        case Add(n, replyTo) => persistAdd(n, replyTo, Done)
+
+        case AddIfLessThan(toAdd, ifLessThan, replyTo) =>
+          if (state.count >= ifLessThan) {
+            context.log.info("Rejecting AddIfLessThan as count = {}", state.count)
+            Effect.none[State].thenRun(_ => replyTo ! false)
+          } else persistAdd(toAdd, replyTo, true)
+
+        case AddWhenAtLeast(toAdd, whenAtLeast, replyTo) =>
+          if (state.count < whenAtLeast) Effect.stash()
+          else persistAdd(toAdd, replyTo, Done)
+
+        case NotifyIfAtLeast(n, notifyTo, replyTo) =>
+          if (state.count >= n) {
+            Effect.none[State].thenRun { _ =>
+              notifyTo ! Done
+              replyTo ! true
+            }
+          } else if (state.notifyAfter.isDefinedAt(n)) {
+            Effect.none[State].thenRun(_ => replyTo ! false)
+          } else {
+            Effect.persist(state.addObserver(n, notifyTo)).thenRun(_ => replyTo ! true)
+          }
+
+        case GetRevisionNumber(replyTo) =>
+          Effect.none[State].thenRun(_ => replyTo ! DurableStateBehavior.lastSequenceNumber(context))
+      }
+    }
+  }
+}
+
+class UnpersistentDurableStateSpec extends AnyWordSpec with Matchers {
+  import akka.actor.testkit.typed.scaladsl._
+  import org.slf4j.event.Level
+
+  import UnpersistentDurableStateSpec._
+
+  "Unpersistent DurableStateBehavior" must {
+    "generate a fail-fast behavior from a non-DurableStateBehavior" in {
+      val notDurableState =
+        Behaviors.receive[Any] { (context, msg) =>
+          context.log.info("Got message {}", msg)
+          Behaviors.same
+        }
+
+      val (unpersistent, probe) = UnpersistentBehavior.fromDurableState[Any, Any](notDurableState)
+      an[AssertionError] shouldBe thrownBy { BehaviorTestKit(unpersistent) }
+      assert(!probe.hasEffects, "should be no persistence effects")
+    }
+
+    "generate a Behavior from a DurableStateBehavior and process RecoveryCompleted" in {
+      import BehaviorUnderTest._
+
+      val recoveryDone = TestInbox[Done]()
+      val behavior = BehaviorUnderTest("test-1", recoveryDone.ref)
+
+      val (unpersistent, probe) = UnpersistentBehavior.fromDurableState[Command, State](behavior)
```

Comment:
I know it is just testkit and not production API but returning tuples is not great DX, let's make a result type for Scala as well instead.

Reply:
So with a result type patterned on the Java API, you end up with code like

```scala
val unpersistent = UnpersistentBehavior.fromDurableState[Command, State](behavior)
val probe = unpersistent.stateProbe
val testkit = unpersistent.behaviorTestKit
```

or with an extractor, something like the tuple return

```scala
val UnpersistentBehavior.DurableState(testkit, probe) =
  UnpersistentBehavior.fromDurableState[Command, State](behavior)
```

OTOH, in the Scala API, maybe it makes the most sense to have the result type take a `(BehaviorTestKit[Command], PersistenceProbe[Event], PersistenceProbe[State]) => Unit` (or otherwise for durable state) in addition to the accessors?

So it would be like

```scala
UnpersistenceBehavior.fromDurableState[Command, State](behavior) { (testkit, probe) =>
  // body of test
}
```

Reply:
That does make the Java API rather different from the Scala API, but I don't get the sense that that sort of resource-like API is at all idiomatic in Java.

Reply:
A case class so that it is possible to use the unapply if you like, or the object with named fields if you rater like that sounds good to me (better than 3-tuple).